File: intern/1894.find-the-student-that-will-replace-the-chalk.js
/*
 * @lc app=leetcode id=1894 lang=javascript
 *
 * [1894] Find the Student that Will Replace the Chalk
 */

// @lc code=start
/**
 * @param {number[]} chalk
 * @param {number} k
 * @return {number}
 */
const chalkReplacer = function (chalk, k) {
    const chalkSum = chalk.reduce((prev, current) => prev + current);
    k = k % chalkSum;

    const STUDENT_AMOUNT = chalk.length;
    for (let i = 0; i < STUDENT_AMOUNT; i++) {
        k -= chalk[i];
        if (k < 0) {
            return i;
        }
    }
};
// @lc code=end

File: prev-2024/1894.find-the-student-that-will-replace-the-chalk.cpp
/*
 * @lc app=leetcode id=1894 lang=cpp
 *
 * [1894] Find the Student that Will Replace the Chalk
 */
#include <numeric>
#include <vector>

// @lc code=start
class Solution {
 public:
  int chalkReplacer(std::vector<int>& chalk, int k) {
    const int kStudentNumber = chalk.size();

    int chalkSum = 0;
    for (int i = 0; i < kStudentNumber; i++) {
      chalkSum += chalk[i];
      if (chalkSum > k) {
        return i;
      }
    }

    k %= chalkSum;

    for (int i = 0; i < kStudentNumber; i++) {
      const int neededChalk = chalk[i];
      if (neededChalk > k) {
        return i;
      }
      k -= neededChalk;
    }

    return -1;
  }
};
// @lc code=end
